apps/arweave/src/ar_chunk_cache.erl

```diff
@@ -0,0 +1,495 @@
+    chunk_cache_groups_queue = queue:filter(
+      fun(GroupId0) -> GroupId0 =/= GroupId end,
+      Cache0#ar_chunk_cache.chunk_cache_groups_queue
+    )
+  }.
+
+
+
+-spec group_exists(GroupId :: term(), Cache0 :: #ar_chunk_cache{}) ->
+  Exists :: boolean().
+
+group_exists(GroupId, Cache0) ->
+  maps:is_key(GroupId, Cache0#ar_chunk_cache.chunk_cache_groups).
+
+
+
+-spec get_groups(Cache0 :: #ar_chunk_cache{}) ->
+  Groups :: [term()].
+
+get_groups(Cache0) ->
+  queue:to_list(Cache0#ar_chunk_cache.chunk_cache_groups_queue).
+
+
+
+-spec add_chunk_to_existing_group(
+  GroupId :: term(),
+  ChunkId :: term(),
+  Chunk :: binary() | ?CACHE_VALUE(binary(), #{}),
+  Cache0 :: #ar_chunk_cache{}
+) ->
+  {ok, Cache1 :: #ar_chunk_cache{}} | {error, Reason :: term()}.
+
+add_chunk_to_existing_group(GroupId, ChunkId, ?CACHE_VALUE(Chunk, ChunkMeta), Cache0) ->
+  case (byte_size(Chunk) + cache_size(Cache0)) >  Cache0#ar_chunk_cache.chunk_cache_limit_bytes of
+    true when Cache0#ar_chunk_cache.chunk_cache_limit_bytes =/= 0 -> {error, cache_limit_exceeded};
+    _ -> map_chunk_cache_group(GroupId, add_chunk_map_fun(ChunkId, Chunk, ChunkMeta), Cache0)
+  end;
+
+add_chunk_to_existing_group(GroupId, ChunkId, Chunk, Cache0) ->
+  add_chunk_to_existing_group(GroupId, ChunkId, ?CACHE_VALUE(Chunk, #{}), Cache0).
+
+
+
+-spec add_chunk(
+  GroupId :: term(),
+  ChunkId :: term(),
+  Chunk :: binary() | {binary(), #{}},
+  Cache0 :: #ar_chunk_cache{}
+) ->
+  {ok, Cache1 :: #ar_chunk_cache{}} | {error, Reason :: term()}.
+
+add_chunk(GroupId, ChunkId, ?CACHE_VALUE(Chunk, ChunkMeta), Cache0) ->
+  case (byte_size(Chunk) + cache_size(Cache0)) >  Cache0#ar_chunk_cache.chunk_cache_limit_bytes of
+    true when Cache0#ar_chunk_cache.chunk_cache_limit_bytes =/= 0 -> {error, cache_limit_exceeded};
+    _ -> map_chunk_cache_group(GroupId, add_chunk_map_fun(ChunkId, Chunk, ChunkMeta), Cache0, true)
+  end;
+
+add_chunk(GroupId, ChunkId, Chunk, Cache0) ->
+  add_chunk(GroupId, ChunkId, ?CACHE_VALUE(Chunk, #{}), Cache0).
+
+
+
+-spec take_chunk(GroupId :: term(), ChunkId :: term(), Cache0 :: #ar_chunk_cache{}) ->
+  {ok, Chunk :: binary(), Cache1 :: #ar_chunk_cache{}} | {error, Reason :: term()}.
+
+take_chunk(GroupId, ChunkId, Cache0) ->
+  map_chunk_cache_group(GroupId, fun(#ar_chunk_cache_group{
+    chunk_cache = ChunkCache0,
+    chunk_cache_size_bytes = ChunkCacheSize0
+  } = Group0) ->
+    case maps:take(ChunkId, ChunkCache0) of
+      {?CACHE_VALUE(Chunk, _Meta) = RetVal, ChunkCache1} -> {ok, RetVal, Group0#ar_chunk_cache_group{
+        chunk_cache = ChunkCache1,
+        chunk_cache_size_bytes = ChunkCacheSize0 - byte_size(Chunk)
+      }};
+      error -> {error, chunk_not_found}
+    end
+  end, Cache0).
+
+
+
+-spec drop_chunk(GroupId :: term(), ChunkId :: term(), Cache0 :: #ar_chunk_cache{}) ->
+  {ok, Cache1 :: #ar_chunk_cache{}} | {error, Reason :: term()}.
+
+drop_chunk(GroupId, ChunkId, Cache0) ->
```

Comment:
I don't think this is used anywhere (maybe used to be?)

Reply:
No, it isn't, this function is here for API consistency.